[JIT] Fuser match on schemas not node kind

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44138 [JIT] Remove references to no longer generated _tanh_backward and _sigmoid_backward
* #44137 [TensorExpr fuser] Guard nodes that have tensor output properties determined by non-tensor inputs
* **#44083 [JIT] Fuser match on schemas not node kind**

Match on the complete schema of a node instead of its node kind when deciding to fuse it. Previously we matched on node kind, which could fail with something like `aten::add(int, int)` and if a new overload was added to an op without corresponding NNC support we would fuse it. 

Follow ups are:
 - bail when an output tensor type isnt uniquely determined by the input types (e.g. aten::add and the second input could be either a float or an int)
- remove NNC lowering for _tanh_backward & _sigmoid_backward
- Validate that we support all of the overloads here. I optimistically added ops that included Tensors, it's possible that we do not support every overload here. This isn't a regression, and this PR is at least improving our failures in that regard. 

I can do any of these as part of this PR if desired, but there are a number of failures people have run into that this PR fixes so I think it would be good to land this sooner than later.

Differential Revision: [D23503704](https://our.internmc.facebook.com/intern/diff/D23503704)

Commit:
Match on schemas not node kind

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -31,90 +31,113 @@ bool isSupported(Node* node) {
-    case aten::frac:
-    // TODO: uncomment once we can handle rand+broadcasts
-    // case aten::rand_like:
-    case aten::_sigmoid_backward:
-    case aten::_tanh_backward:
```
These were the only two ops that I removed - I can reinclude them if wished. I dont think they are generated anywhere. If you look at the sigmoid and tanh backward they are composed of other ops: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/symbolic_script.cpp#L786
https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/symbolic_script.cpp#L164

The corresponding NNC code should probably also be removed but i'm leaving that for a follow up

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -31,90 +31,113 @@ bool isSupported(Node* node) {
-  // TODO:
+
+  static const OperatorSet supported_operator_set{
+      "aten::add.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor",
```
The operator order & comments are preserved from the original set of node kinds so it should be easy to review this

Commit:
Update on "[JIT] Match on schemas not node kind"




[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -31,90 +31,113 @@ bool isSupported(Node* node) {
+      "aten::expm1(Tensor self) -> Tensor",
+      "aten::unsqueeze(Tensor(a) self, int dim) -> Tensor(a)",
+      "aten::frac(Tensor self) -> Tensor",
+      // TODO: uncomment once we can handle rand+broadcasts
```
Nit: uncomment will no longer work here :D

ZolotukhinM reviewed (approved):
Looks good to me! We will probably need a similar change in NNC shape inference.

bertmaher reviewed (approved):
awesomesauce

Commit:
Update on "[JIT] Fuser match on schemas not node kind"


Match on the complete schema of a node instead of its node kind when deciding to fuse it. Previously we matched on node kind, which could fail with something like `aten::add(int, int)` and if a new overload was added to an op without corresponding NNC support we would fuse it. 

Follow ups are:
 - bail when an output tensor type isnt uniquely determined by the input types (e.g. aten::add and the second input could be either a float or an int)
- remove NNC lowering for _tanh_backward & _sigmoid_backward
- Validate that we support all of the overloads here. I optimistically added ops that included Tensors, it's possible that we do not support every overload here. This isn't a regression, and this PR is at least improving our failures in that regard. 

I can do any of these as part of this PR if desired, but there are a number of failures people have run into that this PR fixes so I think it would be good to land this sooner than later.

[ghstack-poisoned]

Commit:
Update on "[JIT] Fuser match on schemas not node kind"


Match on the complete schema of a node instead of its node kind when deciding to fuse it. Previously we matched on node kind, which could fail with something like `aten::add(int, int)` and if a new overload was added to an op without corresponding NNC support we would fuse it. 

Follow ups are:
 - bail when an output tensor type isnt uniquely determined by the input types (e.g. aten::add and the second input could be either a float or an int)
- remove NNC lowering for _tanh_backward & _sigmoid_backward
- Validate that we support all of the overloads here. I optimistically added ops that included Tensors, it's possible that we do not support every overload here. This isn't a regression, and this PR is at least improving our failures in that regard. 

I can do any of these as part of this PR if desired, but there are a number of failures people have run into that this PR fixes so I think it would be good to land this sooner than later.

Differential Revision: [D23503704](https://our.internmc.facebook.com/intern/diff/D23503704)

[ghstack-poisoned]
